### PR TITLE
Infrastructure/3072 add use exhaustive deps to es lint

### PR DESCRIPTION
**Pull request by danielgent**
## Summary

Addresses issue #3072

## Checklist

- [ ] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

File: .eslintrc.json [hunk 1]
```diff
@@ -2,7 +2,8 @@
   "extends": [
     "plugin:@wordpress/eslint-plugin/recommended-with-formatting",
     "plugin:jest/recommended",
-    "plugin:sitekit/main"
+    "plugin:sitekit/main",
+    "plugin:react-hooks/recommended"
```
Comment (danielgent) [inline .eslintrc.json:6]: The IB says to only enable one of `react-hook`'s two rules, *but* enabling `react-hooks/rules-of-hooks` didn't generate a single warning (Eugene was worried here => https://github.com/google/site-kit-wp/issues/3072#issuecomment-817684043 )

`react-hooks/rules-of-hooks` gives great warnings for people who are learning hooks

File: assets/js/components/ActivateModuleCTA.js [hunk 2]
```diff
@@ -62,7 +62,7 @@ const ActivateModuleCTA = ( { moduleSlug, title, description } ) => {
 				type: 'win-error',
 			} );
 		}
-	}, [ activateModule ] );
+	}, [ activateModule, navigateTo, moduleSlug ] );
```
Comment (danielgent) [inline assets/js/components/ActivateModuleCTA.js:65]: most of this PR are `useCallback` dependency arrays being populated with all the dependencies that eslint tells us

File: assets/js/components/GoogleChart.js [hunk 3]
```diff
@@ -169,7 +171,7 @@ export default function GoogleChart( props ) {
 
 			clearInterval( interval );
 		};
-	}, [] );
+	} );
```
Comment (danielgent) [inline assets/js/components/GoogleChart.js:174]: I've assumed that any `useEffect` with an empty dependency array is meant to only run on mount (as opposed to being an error!)

File: assets/js/components/GoogleChart.js [hunk 4]
```diff
@@ -88,7 +89,7 @@ export default function GoogleChart( props ) {
 
 			setChart( googleChart );
 		}
-	}, [ loading, !! chartRef.current, visualizationLoaded, !! chart ] );
+	}, [ loading, visualizationLoaded, chart, chartID, chartType, onReady ] );
```
Comment (danielgent) [inline assets/js/components/GoogleChart.js:92]: There are some dependency arrays in the code which have values that don't look like React best practices, and eslint gives all sorts of warnings about these now (not having complex values in the dependencies array, not having unused values), so I'm following eslint's recommendations

File: assets/js/components/GoogleChart.js [hunk 5]
```diff
@@ -126,13 +127,14 @@ export default function GoogleChart( props ) {
 		};
 	}, [
 		chart,
-		JSON.stringify( data ),
-		JSON.stringify( options ),
 		selectedStats,
 		singleStat,
+		chartID,
+		data,
+		options,
```
Comment (danielgent) [inline assets/js/components/GoogleChart.js:134]: same with this. `JSON.stringify` creates strings which are immutable and equal in JS if they have the same value, but that is not true with the object references I'm not passing in :thinking: 

The tests still pass but I'm worried in this file especially that I may have changed something (I'm happy to revert this and just disable eslint here)

File: assets/js/components/ResetButton.js [hunk 6]
```diff
@@ -49,11 +49,14 @@ function ResetButton( { children } ) {
 	 * the navigate call starting, we will just set a debounce to keep the spinner for 3 seconds.
 	 */
 	const debouncedSetInProgress = useDebounce( setInProgress, 3000 );
-	const mediatedSetInProgress = ( bool ) => bool ? setInProgress( true ) : debouncedSetInProgress( false );
+	const mediatedSetInProgress = useCallback(
```
Comment (danielgent) [inline assets/js/components/ResetButton.js:52]: this dependency to a callback also has to be a callback (eslint now warns us)

File: assets/js/components/help/HelpMenu.js [hunk 7]
```diff
@@ -66,7 +66,7 @@ function HelpMenu( { children } ) {
 
 	const handleMenuSelected = useCallback( () => {
 		toggleMenu( false );
-	} );
+	}, [] );
```
Comment (danielgent) [inline assets/js/components/help/HelpMenu.js:69]: There's no reason for this to be a callback now technically but as per our tech decision, we want all handlers to use `useCallback` now

Comment (danielgent) [reply assets/js/components/help/HelpMenu.js:69]: `eslint-plugin-react-hooks` knows that `toggleMenu` comes from `useState` and thus does not need to be a dependency. Which is pretty cool

File: assets/js/components/settings/SettingsActiveModule/ConfirmDisconnect.js [hunk 8]
```diff
@@ -73,7 +73,7 @@ export default function ConfirmDisconnect( { slug, handleDialog } ) {
 			clearWebStorage();
 			global.location.assign( adminReauthURL );
 		}
-	}, [ module?.slug ] );
+	}, [ adminReauthURL, deactivateModule, module.forceActive, slug ] );
```
Comment (danielgent) [inline assets/js/components/settings/SettingsActiveModule/ConfirmDisconnect.js:76]: as this is a `useCallback` and not `useEffect` then IMHO we can just follow `eslint-plugin-react-hooks` recommendations (it's not re-running code on any change here, it's just assuring there are no stale closures)

File: assets/js/components/user-input/UserInputQuestionnaire.js [hunk 9]
```diff
@@ -48,9 +48,9 @@ import { Cell, Row } from '../../material-components';
 import { trackEvent } from '../../util';
 const { useSelect, useDispatch } = Data;
 
-export default function UserInputQuestionnaire() {
-	const steps = [ ...USER_INPUT_QUESTIONS_LIST, 'preview' ];
+const steps = [ ...USER_INPUT_QUESTIONS_LIST, 'preview' ];
```
Comment (danielgent) [inline assets/js/components/user-input/UserInputQuestionnaire.js:51]: move out of component so does not become a dynamic dependency

File: assets/js/modules/analytics/components/common/AccountCreate/CountrySelect.js [hunk 10]
```diff
@@ -48,7 +48,7 @@ export default function CountrySelect() {
 				timezone: countriesByCode[ newCountryCode ].defaultTimeZoneId,
 			} );
 		}
-	}, [ setValues ] );
+	}, [ setValues, value ] );
```
Comment (danielgent) [inline assets/js/modules/analytics/components/common/AccountCreate/CountrySelect.js:51]: I don't know how we haven't got bugs from some of these! As here this would have had a stale `value`

File: assets/js/modules/analytics/components/dashboard/DashboardAllTrafficWidget/DimensionTabs.js [hunk 11]
```diff
@@ -45,24 +45,24 @@ import { Select, Option } from '../../../../../material-components';
 import { trackEvent } from '../../../../../util';
 const { useDispatch } = Data;
 
+const tabs = [
```
Comment (danielgent) [inline assets/js/modules/analytics/components/dashboard/DashboardAllTrafficWidget/DimensionTabs.js:48]: extracted out as needs to be static as is a dependency

File: assets/js/modules/adsense/components/dashboard/LegacyAdSenseDashboardWidgetOverview.js [hunk 12]
```diff
@@ -85,9 +85,13 @@ export default function LegacyAdSenseDashboardWidgetOverview( props ) {
...
-		currentDataLoaded && previousDataLoaded, // All reports are fetched.
-		!! currentError || !! previousError, // Whether there is an error or not.
-		JSON.stringify( currentRangeData ),
+		currentDataLoaded,
+		currentError,
+		currentRangeData,
+		handleDataError,
+		handleDataSuccess,
+		previousDataLoaded,
+		previousError,
```
Comment (danielgent) [inline assets/js/modules/adsense/components/dashboard/LegacyAdSenseDashboardWidgetOverview.js:94]: As this is a `useEffect` this could be a cause for concern!

Note that `useEffect` re-runs when a dependency changes **not** when a dependency is truthy (the previous code does look a bit like it's expecting that)

File: assets/js/modules/tagmanager/components/setup/SetupForm.js [hunk 13]
```diff
@@ -61,7 +60,6 @@ export default function SetupForm( { finishSetup } ) {
 	const { setValues } = useDispatch( CORE_FORMS );
 	const { activateModule } = useDispatch( CORE_MODULES );
 	const { submitChanges } = useDispatch( STORE_NAME );
-	const dispatchAnalytics = useDispatch( MODULES_ANALYTICS );
```
Comment (danielgent) [inline assets/js/modules/tagmanager/components/setup/SetupForm.js:64]: :thinking:   I'm assuming this was left around from a previous change. But it's fine to remove

File: assets/js/modules/analytics/components/dashboard/DashboardAllTrafficWidget/UserDimensionsPieChart.js [hunk 14]
```diff
@@ -81,16 +81,18 @@ export default function UserDimensionsPieChart( {
...
 
 		return () => {
-			if ( containerRef.current ) {
-				containerRef.current.removeEventListener( 'click', onTooltipClick );
+			if ( currentContainerRef ) {
+				currentContainerRef.removeEventListener( 'click', onTooltipClick );
 			}
 		};
-	}, [ containerRef.current ] );
+	}, [ ] );
```
Comment (danielgent) [inline assets/js/modules/analytics/components/dashboard/DashboardAllTrafficWidget/UserDimensionsPieChart.js:95]: I'm not sure what was intended before here.  

Changing a ref doesn't re-render so this was just attaching the handler when mounting.  The original div will probably persist because of React's reconciliation algorithm preserving the DOM *but* it could wipe it at any point!

I'm not sure why the handler isn't just assigned straight to the `div` below in the JSX 
e.g. instead of `<div ref={ containerRef }` do `<div onClick={onTooltipClick}`

File: assets/js/components/user-input/UserInputKeywords.js [hunk 15]
```diff
@@ -57,9 +56,6 @@ export default function UserInputKeywords( { slug, max } ) {
 	// https://github.com/google/site-kit-wp/issues/2900#issuecomment-814843972.
 	const [ localValues, setLocalValues ] = useState( values );
 
-	// Need to make sure that dependencies list always has the same number of elements.
-	const dependencies = values.concat( Array( max ) ).slice( 0, max );
```
Comment (danielgent) [inline assets/js/components/user-input/UserInputKeywords.js:61]: I'm not following this. The dependencies array needs to have the dependencies so that the callback doesn't have stale values.

File: assets/js/components/GoogleChart.js (hunk 5 above)
Comment (danielgent) [reply assets/js/components/GoogleChart.js:134]: Note that this is a `useEffect`, the `useEffect` changes are what concern me!

Comment (danielgent) [reply assets/js/components/GoogleChart.js:134]: As we are re-running code (where as `useCallback` is just creating a new function that could be ran. so there are no side-effects)

File: assets/js/modules/tagmanager/hooks/useExistingTagEffect.js [hunk 16]
```diff
@@ -42,5 +42,13 @@ export default function useExistingTagEffect() {
...
-	}, [ hasExistingTag, existingTag, hasExistingTagPermission, existingTagPermission ] );
+	},
+	[
+		hasExistingTag,
+		existingTag,
+		hasExistingTagPermission,
+		existingTagPermission,
+		selectAccount,
+		selectContainerByID,
+	]	);
```
Comment (tofumatt) [inline assets/js/modules/tagmanager/hooks/useExistingTagEffect.js:53]: Surprised ESLint doesn't care about this, but how about:

```suggestion
	}, [
		hasExistingTag,
		existingTag,
		hasExistingTagPermission,
		existingTagPermission,
		selectAccount,
		selectContainerByID,
	] );
```

**Review by tofumatt — CHANGES_REQUESTED** [review]
I'm a bit curious why these rules request usage of `useMount`, because there's nothing special about it: [it's just `useEffect(fn, [])`](https://github.com/streamich/react-use/blob/master/src/useEffectOnce.ts#L4).

`useEffect(fn, [])` isn't an obscure React paradigm or anything, it's a very common hooks approach, so I don't see the need for a new function that obscures what's happening here.

The ACs/IBs don't mention including `useMount` at all, and given it's a wrapper around exactly what we're doing I think we should ignore that rule/revert the change. Arguably it's more boilerplate for many React developers who would think to use `useEffect`. Instead of the empty array argument "boilerplate" it's the different import.

I'd rather we stick with `useEffect` as it's immediately recognisable for any React developer so requires less developer onboarding.

----

The rest of the PR largely looks good—I've made some comments about a few things I think need fixing/tweaking, but overall looks good 👍🏻

File: assets/js/components/GoogleChart.js (hunk 5 above)
Comment (danielgent) [reply assets/js/components/GoogleChart.js:134]: Is that a request @tofumatt ?  :smile:   I'm happy to disable eslint for any existing code

**Comment by tofumatt** [comment]
Looks like this is failing the linters too:

Screenshot 2021-04-23 at 13 54 50

File: assets/js/components/settings/SettingsActiveModule/ConfirmDisconnect.js [hunk 17]
```diff
@@ -16,11 +16,6 @@
  * limitations under the License.
  */
 
-/**
- * External dependencies
- */
-import PropTypes from 'prop-types';
```
Comment (eugene-manuilov) [inline assets/js/components/settings/SettingsActiveModule/ConfirmDisconnect.js:22]: @danielgent why did you delete prop types for this component? Could you, please, add it back?

File: assets/js/components/user-input/UserInputKeywords.js [hunk 18]
```diff
@@ -43,7 +43,6 @@ const { useSelect, useDispatch } = Data;
 
 export default function UserInputKeywords( { slug, max } ) {
 	const keywordsContainer = useRef();
-	const [ canDeleteKeyword, setCanDeleteKeyword ] = useState( false );
```
Comment (eugene-manuilov) [inline assets/js/components/user-input/UserInputKeywords.js:46]: @danielgent why did you delete this?

File: assets/js/components/ResetButton.js [hunk 19]
```diff
@@ -49,11 +49,14 @@ function ResetButton( { children } ) {
...
+		( bool ) => bool ? setInProgress( true ) : debouncedSetInProgress( false ),
+		[ debouncedSetInProgress ]
+	);
```
[the remark this answers is not in the document]
Comment (danielgent) [reply assets/js/components/ResetButton.js]: @tofumatt If we're allowed to make code changes then I inlined `mediatedSetInProgress` right into the `useEffect` (it's not used elsewhere) **AND** I used proper if-else blocks (as per your suggestion)

File: package.json [hunk 20]
```diff
@@ -122,6 +122,7 @@
     "compare-versions": "^3.6.0",
     "dompurify": "^2.0.12",
     "element-closest": "^3.0.2",
+    "eslint-plugin-react-hooks": "^4.2.0",
```
Comment (eugene-manuilov) [inline package.json:125]: @danielgent please, add it as a dev dependency.

**Review by eugene-manuilov — DISMISSED** [review]
Thanks, @danielgent! It looks good in general. Just found a few more issues to address. Could you please check it?

File: assets/js/components/settings/SettingsActiveModule/ConfirmDisconnect.js (hunk 8 above)
Comment (danielgent) [reply assets/js/components/settings/SettingsActiveModule/ConfirmDisconnect.js]: Fixed @tofumatt

File: package.json (hunk 20 above)
Comment (danielgent) [reply package.json]: Ooops @eugene-manuilov ! nice spot! :facepalm: . Fixed :heavy_check_mark:

File: assets/js/components/settings/SettingsActiveModule/ConfirmDisconnect.js (hunk 17 above)
Comment (danielgent) [reply assets/js/components/settings/SettingsActiveModule/ConfirmDisconnect.js]: Apolgies @eugene-manuilov .  I managed to remove them in a merge :thinking:   No idea what happened there! They're added back now :heavy_check_mark:

File: assets/js/components/user-input/UserInputKeywords.js (hunk 18 above)
Comment (danielgent) [reply assets/js/components/user-input/UserInputKeywords.js:46]: @eugene-manuilov This file really confused me. I'm happy to revert it all and just disable eslint for the file :smile: 

This state didn't appear to do anything. `canDeleteKeyword` was only used for unnecessary dependencies.

the react hooks eslint plugin forced me to remove it as a dependency, and then linting forced me to delete the variable due to it being unused!

File: assets/js/components/GoogleChart.js (hunk 5 above)
Comment (danielgent) [reply assets/js/components/GoogleChart.js]: @tofumatt  I reverted this file and just added an eslint ignore for this rule :+1:

**Review by tofumatt — COMMENTED** [review]
I had a bit of a conversion with @danielgent about this and the reasoning for not allowing empty `useEffect` dependency arrays is pretty fair. It adds complexity, but reduces potential for errors in an acceptable way.

So I retract my request for `useMount` to be removed; let's add it back.

File: assets/js/modules/tagmanager/hooks/useExistingTagEffect.js [hunk 21]
```diff
@@ -42,5 +42,12 @@ export default function useExistingTagEffect() {
...
 		} )();
-	}, [ hasExistingTag, existingTag, hasExistingTagPermission, existingTagPermission ] );
+	},	[
+		hasExistingTag,
+		existingTag,
+		hasExistingTagPermission,
+		existingTagPermission,
+		selectAccount,
+		selectContainerByID,
+	]	);
```
Comment (tofumatt) [inline assets/js/modules/tagmanager/hooks/useExistingTagEffect.js:52]: What's with the indentation of the `[` and `]` here?

```suggestion
	}, [
		hasExistingTag,
		existingTag,
		hasExistingTagPermission,
		existingTagPermission,
		selectAccount,
		selectContainerByID,
	] );
```

**Review by tofumatt — CHANGES_REQUESTED** [review]
Overall looks great! One little styling tweak and a few merge conflicts need resolving, but then this should be good-to-merge. 👍🏻

**Review by tofumatt — APPROVED** [review]
Looks good to me, thanks for the work on this! 👍🏻

File: assets/js/modules/analytics/components/dashboard/DashboardAllTrafficWidget/UserDimensionsPieChart.js (hunk 14 above)
Comment (tofumatt) [reply assets/js/modules/analytics/components/dashboard/DashboardAllTrafficWidget/UserDimensionsPieChart.js:95]: Actually, upon further inspection this is a very specific handler for the tooltips that this chart renders, so keeping this here is appropriate.
